### ISSUE TITLE
While syncing with the testnet1 network around block 9938 - thread '<unknown>' has overflowed its stack

### ISSUE THREAD

**Issue by Thireus** [mimblewimble/grin#573]
I would like to report a stack overflow bug which happened while grin server was still syncing with the network around block 9938.

I followed the steps on https://github.com/ignopeverell/grin/blob/master/doc/build.md. Build grin on ARMv7 server with Debian 9.

For some reason after several hours of uptime, the grin server has crashed:

```
thread '<unknown>' has overflowed its stack
fatal runtime error: stack overflow
Aborted
```

Last grin.log:

```
Jan 03 22:28:09.206 DEBG Sumtree extension done.
Jan 03 22:28:09.207 DEBG (Server ID: Port 13414) Mining at Cuckoo16 for at most 90 secs on block a7be852c at difficulty 500.
Jan 03 22:28:09.207 DEBG Starting new sumtree extension.
Jan 03 22:28:09.395 DEBG (Server ID: Port 13414) No solution found after 1 iterations, continuing...
Jan 03 22:28:09.396 DEBG setting pubkey in miner to pubkey from block_fees - BlockFees { fees: 0, height: 9938, key_id: Some(Identifier(b88c8862b83928880a37)) }
Jan 03 22:28:09.396 DEBG in miner loop...
Jan 03 22:28:09.397 TRCE key_id: Some(Identifier(b88c8862b83928880a37))
Jan 03 22:28:09.878 DEBG chain_store: get_output_pos: Commitment(097193c5eddd79daaf48f1007a66991998a82992b14bf27b9dd5aca0f12a1b87a9), Err(NotFoundErr)
Jan 03 22:28:09.879 DEBG sumtree: get_output_pos (idx): Commitment(097193c5eddd79daaf48f1007a66991998a82992b14bf27b9dd5aca0f12a1b87a9), Err(NotFoundErr)
Jan 03 22:28:09.879 TRCE Hashing with additional data
```

I'm not sure what else I can do to debug this issue. I've restarted the server in the meantime to continue syncing with the testnet1 network.

**Comment by antiochp** [comment]
Hmm - I think I've seen this a couple of times as well. Doesn't happen all the time.
I believe there is a recursive call somewhere (maybe handling a really old fork before we ban a peer?) that recurses over the full block chain (or header chain) and the chain is too big to do this safely via recursion.

Will dig into this further if I get chance.

Thanks for the report!

**Comment by Jiloc** [comment]
Just happened to me also. Ubuntu 17.10, branch testnet1. I was following the Basic Execution tutorial. Here the last log output.

```
Jan 03 23:38:03.746 DEBG Starting new sumtree extension.
Jan 03 23:38:03.752 DEBG (Server ID: Port 13414) Mining at Cuckoo16 for 90 secs (will wait for last solution) on block 1e90768d at difficulty 287.
Jan 03 23:38:03.754 DEBG (Server ID: Port 13414) No solution found after 1 iterations, continuing...
Jan 03 23:38:03.754 DEBG setting pubkey in miner to pubkey from block_fees - BlockFees { fees: 0, height: 4473, key_id: Some(Identifier(a4197ce0c1b847778f95)) }
Jan 03 23:38:03.754 DEBG in miner loop...
Jan 03 23:38:03.754 TRCE key_id: Some(Identifier(a4197ce0c1b847778f95))
Jan 03 23:38:03.766 INFO Acquiring wallet lock ...
Jan 03 23:38:03.766 DEBG Attempting to acquire wallet lock
Jan 03 23:38:03.767 INFO ... released wallet lock
Jan 03 23:38:03.767 DEBG Received coinbase and built candidate output - Identifier(a2c23bf7910dafc21242), Identifier(a4197ce0c1b847778f95), 4
Jan 03 23:38:03.768 DEBG block_fees - BlockFees { fees: 0, height: 4476, key_id: Some(Identifier(a4197ce0c1b847778f95)) }
Jan 03 23:38:03.768 DEBG block_fees updated - BlockFees { fees: 0, height: 4476, key_id: Some(Identifier(a4197ce0c1b847778f95)) }
Jan 03 23:38:03.768 TRCE Derived Key by key_id: a4197ce0c1b847778f95
Jan 03 23:38:03.768 TRCE ... Derived Key (cache hit) key_id: a4197ce0c1b847778f95, derivation: 4
Jan 03 23:38:03.768 TRCE Derived Key (fast) by derivation: 4
Jan 03 23:38:03.773 TRCE Derived Key by key_id: a4197ce0c1b847778f95
Jan 03 23:38:03.773 TRCE ... Derived Key (cache hit) key_id: a4197ce0c1b847778f95, derivation: 4
Jan 03 23:38:03.773 TRCE Derived Key (fast) by derivation: 4
Jan 03 23:38:03.774 TRCE Block reward - Pedersen Commit is: Commitment(096e4e78a92246d8e314f8ce6e8cfbfd045b833ecc4ad4a3074438b49b35042dee), Switch Commit is: Commitment(099018772bf7de72f12465f938d5deddd446f9664b19bd811da469d4541414168d)
Jan 03 23:38:03.774 TRCE Block reward - Switch Commit Hash is: SwitchCommitHash { hash: [126, 102, 9, 120, 190, 170, 121, 55, 117, 78, 127, 113, 40, 115, 59, 148, 46, 165, 197, 73] }
Jan 03 23:38:03.774 TRCE Derived Key by key_id: a4197ce0c1b847778f95
Jan 03 23:38:03.774 TRCE ... Derived Key (cache hit) key_id: a4197ce0c1b847778f95, derivation: 4
Jan 03 23:38:03.775 TRCE Derived Key (fast) by derivation: 4
Jan 03 23:38:03.900 TRCE Derived Key by key_id: a4197ce0c1b847778f95
Jan 03 23:38:03.900 TRCE ... Derived Key (cache hit) key_id: a4197ce0c1b847778f95, derivation: 4
Jan 03 23:38:03.901 TRCE Derived Key (fast) by derivation: 4
Jan 03 23:38:03.923 DEBG block_fees here: BlockFees { fees: 0, height: 4476, key_id: Some(Identifier(a4197ce0c1b847778f95)) }
Jan 03 23:38:03.923 DEBG (Server ID: Port 13414) Built new block with 0 inputs and 1 outputs, difficulty: 294
Jan 03 23:38:04.027 DEBG chain_store: get_output_pos: Commitment(090f2bfd1b228ff522c92006cb7195b316eab8b880320c9f04d341f010c85ff45d), Err(NotFoundErr)
Jan 03 23:38:04.027 DEBG sumtree: get_output_pos (idx): Commitment(090f2bfd1b228ff522c92006cb7195b316eab8b880320c9f04d341f010c85ff45d), Err(NotFoundErr)
Jan 03 23:38:04.027 TRCE Hashing with additional data
```

**Comment by CodeSlip** [comment]
Received the same issue - any help would be awesome


Jan 07 19:09:10.645 WARN Grin server started.

thread '<unknown>' has overflowed its stack
fatal runtime error: stack overflow
Aborted (core dumped)

**Comment by yeastplume** [comment]
Looking into cause it happens with regularity on a node I have (with only 2GB ram) backtrace is:

```
    frame #14: 0x0000555556745345 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff6046168, opts=(bits = 8)) at chain.rs:213
    frame #15: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff6047be0) at chain.rs:335
    frame #16: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff6047a98, opts=(bits = 8)) at chain.rs:231
    frame #17: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff6049510) at chain.rs:335
    frame #18: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff60493c8, opts=(bits = 8)) at chain.rs:231
    frame #19: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff604ae40) at chain.rs:335
    frame #20: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff604acf8, opts=(bits = 8)) at chain.rs:231
    frame #21: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff604c770) at chain.rs:335                                                                                                                                frame #22: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff604c628, opts=(bits = 8)) at chain.rs:231
    frame #23: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff604e0a0) at chain.rs:335                                                                                                                                frame #24: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff604df58, opts=(bits = 8)) at chain.rs:231
    frame #25: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff604f9d0) at chain.rs:335                                                                                                                                frame #26: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff604f888, opts=(bits = 8)) at chain.rs:231
    frame #27: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff6051300) at chain.rs:335
    frame #28: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff60511b8, opts=(bits = 8)) at chain.rs:231                                                                                                           frame #29: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff6052c30) at chain.rs:335
    frame #30: 0x0000555556745682 grin`grin_chain::chain::{{impl}}::process_block(self=0x00007ffff67f12d0, b=Block @ 0x00007ffff6052ae8, opts=(bits = 8)) at chain.rs:231                                                                                                           frame #31: 0x00005555567472a3 grin`grin_chain::chain::{{impl}}::check_orphans(self=0x00007ffff67f12d0, block=0x00007ffff6054560) at chain.rs:335
```

**Comment by yeastplume** [comment]
Fixed by #605 / #606 .. will merge once a bit of a review is doen
